### DISCUSSION TITLE
Materialize POIs from Notes or external source / Moderated Import

### DISCUSSION THREAD

**Discussion by amenk** [streetcomplete/StreetComplete#4019] [Ideas]
Background: We want to import/moderate ~ 36.000 POIs in Addis Ababa, Ethiopia which we have currently in an external .osm file. 
The data is a bit old, so it needs to be checked on the ground. Still the data is considered valuable by members of the local community.

The import mailing list thread is here: https://lists.openstreetmap.org/pipermail/imports/2022-May/006844.html

Being a big fan of StreetComplete for it's easy of use, I am wondering if the app could support us here to efficiently verify this data on the ground.

Ideas we had so far:

1. Just import the data after some basic cleaning, check_date / survey_data years in the past -> would this trigger the "is this still existing?" quest? -> drawback: Potentially non-existing POIs need to be imported first to the OSM database, causing confusing -> I do not like this approach.
2. Import the data but with some special tagging prefix, so it does not confuse others but is still in the main OSM database (imported:shop=kiosk) -> create a custom Quest which can then remove the prefix after the POI was checked on the ground
3. Create some kind of structured OSM notes, with the tags and design a custom quest "Is this imported feature still existing yes/no/improve" -> then the POI would be added from the OSM notes
4. Design and integrate some external API where the quests are pulled from and create POIs on OSM if necessary - add a custom quest for this.
5. Design and integrate some general purpose extendable API which can be used for this purpose and other purposes as well?
6. Use another solution, for example MapComplete

The discussion in https://github.com/streetcomplete/StreetComplete/discussions/2477 seems a bit similar.

What are your thought's?

I am not a good Java / Kotlin programmer and I also understand that we might not be able to add a very custom Quest just for Addis Ababa to the mainline StreetComplete app. That is what 5. is about.

**Comment by FloEdelmann** [comment]
For reference, also #3738 is a similar proposal.

**Comment by mnalis** [comment]
As for the possible code, @Helium314 [StreetComplete fork](https://github.com/Helium314/StreetComplete/) contains experimental support for `External quests` (although it doesn't allow user to do much with them, yet)

**Reply by Helium314** [reply]
The Osmose quest might be a better example, as this is based on network connections and not on local files.

I found that I use the external quests less than expected, so the experimental status will stay for longer...
And I expect using large files will lead to pretty bad performance (not sure if 36000 entries would be too much)

**Reply by amenk** [reply]
@Helium314 I tried the Osmose quest in your fork and did not see any quests being loaded, even in Paris where are a lot displayed on the http://osmose.openstreetmap.fr/ - am I doing something wrong?

Anyways, Osmose looks generally very promising.

What I did not find out: Is Osmose / the Osmose quest able to create new POIs? Or can Osmose tasks only modify existing data?

**Reply by amenk** [reply]
I found this:

https://wiki.openstreetmap.org/wiki/Osmose#Issues_file_format

      

So creating new objects is supported in general by Osmose :+1:

**Reply by amenk** [reply]
-> https://github.com/Helium314/StreetComplete/issues/384 & https://github.com/Helium314/StreetComplete/issues/383

**Comment by mnalis** [comment]
- I dislike **1**. - **3.**, because they pollute the map with lots of "junk", on the hopes that mappers will have enough will and interest to clean it up eventually - which might happen to be hoping in vain (and leave OSM cluttered with useless things, causing endless waste of time and resources for all future mappers)

- Solution like **6.** (MapComplete) sounds like a best bet which can be done with current technologies, but I haven't had much happiness with trying to use MapComplete on mobile. Maybe it's just me, though.

That being said, I would very much **love** StreetComplete (or something with similarly nice to use UI) to implement something like **4.** / **5.**, but it would needs some effort to define it in a way to to be fully functional. (SC UI would need just one option in preferences - URL to server with BBOX placeholders)

The good part is that the server could prepare all the quest so can be as easy (e..g static JSON file if smaller area) or as complex (dynamic scripts consulting SQL database etc) as one wants to present the quests to the user.

The problematic part is what the user can do when they're presented with the quest? Sure, they can leave a note (which requires further manual work, and thus has many of the same problems as **3.** -- which is what mentioned @Helium314 fork does by giving a list of e.g. `NODE,2345678,suspicious shop type`), but is there a way for them to contribute directly? If there was, that would be a great way to contribute by doing external custom quests. But it would need to:
- allow either `type`+`id` for existing elements, or `lat`+`lon` (or even whole polygon definition) for nonexistent ones (like in this case)
- send from server to app not only the `question`, but also the `quest explanation`, as well as all possible answers (so it would better be JSON than CSV). It allows only for text-only quests, but that is good enough while keeping complexity low.
- then there is a question what might be done with that user-provided answer?
  -  One way is that app would simply send the answer to the server, and let server handle it (which would make a server much more complex, and also it would likely be problematic as whose account to use to submit changes (either users would have to give server access to their accounts using OAuth, or all changes would be attributed to one account, which is also not nice). Possible advantage would be that android app would not to be further changed in the future no matter how much complexity is increased (e.g. relation support added etc).
  - another way is that the server would send not only answers, but some kind of instructions what to do when that answer is selected.  Which is doable in simpler cases (e.g. `CREATE NODE` at `lat, lon` and/or `ADD/MODIFY/DELETE` `tag=value` on NODE/WAY `id`, but it might also be more complicated if the polygon needs to be created for example) -- but something like this would be much more usable, as it would be very flexible, and user could upload their own data directly to the OSM map using existing interfaces. This would also require that the server sends changeset comments/tags among other information. I would love to see something like this.
